File: cwh/8.js
// 给你两个单链表的头节点 headA 和 headB ，请你找出并返回两个单链表相交的起始节点。如果两个链表不存在相交节点，返回 null 。

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */


function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}


// var getIntersectionNode = function(headA, headB) {
//     let arr1 = [];
//     let arr2 = [];

//     let current1 = headA;
//     let tag = headA;
//     let current2 = headB;

//     while(current1){
//         arr1.push(current1.val);
//         current1 = current1.next;
//     }

//     while(current2){
//         arr2.push(current2.val);
//         current2 = current2.next;
//     }

//     len = 0;
//     len1 = arr1.length;
//     len2 = arr2.length;

//     if(len1 > len2){
//         len = len1;
//     }else{
//         len = len2;
//     }

//     for(let i=0;i<len;i++){
//         if(arr1[i]==arr2[i]){
//             while(tag){
//                 if(tag.val == arr1[i]){
//                     return tag;
//                 }
//                 tag = tag.next;
//             }
//         }
//     }
//     return null;
// };

var getIntersectionNode = function(headA, headB) {
    if (!headA || !headB) return null; // 若任一链表为空，直接返回null
    
    let pA = headA, pB = headB; // 定义两个指针，分别从两个链表头出发
    while (pA !== pB) { // 当两指针未相遇时，继续循环
        // 指针走完当前链表后，切换到另一条链表的头部继续走
        pA = pA ? pA.next : headB; 
        pB = pB ? pB.next : headA;
    }
    return pA; // 相遇时，pA和pB要么是相交节点，要么都是null
};

function a(headA,headB){

    if(!headA || !headB) return null;

    let p1 = headA;
    let p2 = headB;

    while(p1!==p2){
        p1 = p1 ? p1.next : headB;
        p2 = p2 ? p2.next : headA
    }

    return p1;
}

// 1. 创建一个示例链表：1 -> 2 -> 3 -> null
const node3 = new ListNode(3);       // 最后一个节点（next默认null）
const node2 = new ListNode(2, node3); // 第二个节点，next指向node3
const node1 = new ListNode(1, node2); // 第一个节点（头节点），next指向node2

const node4 = new ListNode(4,node2)

console.log(a(node1,node4))

File: yy/8.py
给你一个链表，两两交换其中相邻的节点，并返回交换后链表的头节点。你必须在不修改节点内部的值的情况下完成本题（即，只能进行节点交换）。

示例 1：
输入：head = [1,2,3,4]
输出：[2,1,4,3]

示例 2：
输入：head = []
输出：[]

示例 3：
输入：head = [1]
输出：[1]

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #创建虚拟头节点，简化头节点处理
        dummy=ListNode(0)
        dummy.next=head
        prev=dummy#前一个结点的指针
        while prev.next and prev.next.next:
            node1=prev.next
            node2=prev.next.next
            #交换
            prev.next=node2
            node1.next=node2.next
            node2.next=node1

            prev=node1

        return dummy.next